File: daq/slc/dqm/src/Graph1.cc
#include "daq/slc/dqm/Graph1.h"

using namespace Belle2;

Graph1::Graph1() throw()
  : Histo() {}

Graph1::Graph1(const Graph1& h) throw()
  : Histo(h) {}

Graph1::Graph1(std::string name, std::string title,
               int nbinx, double xmin, double xmax,
               double ymin, double ymax) throw()
  : Histo(name, title)
{
  m_axis_x.set(nbinx, xmin, xmax);
  m_axis_y.set(1, ymin, ymax);
}

Graph1::~Graph1() throw() {}

double Graph1::getPointX(int n) const throw()
{
  return m_data->get(n);
}

double Graph1::getPointY(int n) const throw()
{
  return m_data->get(m_data->size() / 2 + n);
}

void Graph1::setPointX(int n, double data) throw()
{
  m_data->set(n, data);
}

void Graph1::setPointY(int n, double data) throw()
{
  m_data->set(m_data->size() / 2 + n, data);
}

double Graph1::getMaximum() const throw()
{
  return m_axis_y.getMax();
}

double Graph1::getMinimum() const throw()
{
  return m_axis_y.getMin();
}

void Graph1::setMaximum(double data) throw()
{
  if (m_axis_y.isFixedMax()) m_axis_y.setMax(data);
}

void Graph1::setMinimum(double data) throw()
{
  if (m_axis_y.isFixedMin()) m_axis_y.setMin(data);
}

void Graph1::fixMaximum(double data) throw()
{
  m_axis_y.fixMax(true);
  m_axis_y.setMax(data);
}

void Graph1::fixMinimum(double data) throw()
{
  m_axis_y.fixMin(true);
  m_axis_y.setMin(data);
}

void Graph1::readConfig(Reader& reader) throw(IOException)
{
  Histo::readConfig(reader);
  m_axis_x.readObject(reader);
  m_axis_y.readObject(reader);
  m_data->resize(m_axis_x.getNbins() * 2);
}

void Graph1::writeConfig(Writer& writer) const throw(IOException)
{
  Histo::writeConfig(writer);
  m_axis_x.writeObject(writer);
  m_axis_y.writeObject(writer);
}

void Graph1::reset() throw()
{
  setUpdateId(0);
  m_data->setAll(0);
}

File: daq/slc/extra/javalib/b2daq/src/b2daq/dqm/core/Graph1.java
package b2daq.dqm.core;

import b2daq.core.Reader;
import b2daq.core.Writer;
import b2daq.dqm.graphics.GGraph1;
import java.io.IOException;
import javafx.scene.paint.Color;

public abstract class Graph1 extends Histo {

    public Graph1() {
        super();
        setLineColor(Color.RED);
        setFontSize(0.5);
        drawer = new GGraph1(null, this);
    }

    public Graph1(String name, String title, int nbinx, double xmin, double xmax,
            double ymin, double ymax) {
        super(name, title);
        setLineColor(Color.RED);
        setFontSize(0.5);
        axisX.setRange(nbinx, xmin, xmax);
        axisY.setRange(1, ymin, ymax);
        drawer = new GGraph1(null, this);
    }

    @Override
    public void fixMaximum(double data, boolean fix) {
        axisY.setMax(data);
        axisY.setFixedMax(fix);
    }

    @Override
    public void fixMinimum(double data, boolean fix) {
        axisY.setMin(data);
        axisY.setFixedMin(fix);
    }

    @Override
    public int getDim() {
        return 1;
    }

    @Override
    public double getMaximum() {
        return axisY.getMax();
    }

    @Override
    public double getMinimum() {
        return axisY.getMin();
    }

    @Override
    public void setMaximum(double data) {
        if (!axisY.isFixedMax()) {
            axisY.setMax(data);
        }
    }

    @Override
    public void setMinimum(double data) {
        if (!axisY.isFixedMin()) {
            axisY.setMin(data);
        }
    }

    @Override
    public double getPointX(int n) {
        return data.get(n);
    }

    @Override
    public double getPointY(int n) {
        return data.get(data.length() / 2 + n);
    }

    @Override
    public void setPointX(int n, double data) {
        this.data.set(n, data);
    }

    @Override
    public void setPointY(int n, double data) {
        this.data.set(this.data.length() / 2 + n, data);
    }

    @Override
    public void setMaxAndMin() {
        if (!axisY.isFixedMax() || !axisY.isFixedMin()) {
            double data, min = getMinimum(), max = getMaximum();
            for (int nx = 0; nx < axisX.getNbins(); nx++) {
                data = getPointY(nx);
                if (data * 1.05 > max) {
                    max = data * 1.05;
                }
                if (data * 1.05 < min) {
                    min = data * 1.05;
                }
            }
            if (getMaximum() < max && getMinimum() > min) {
                setMaximum((max + min) / 2. + (max - min) / 2. * 1.05);
                setMinimum((max + min) / 2. - (max - min) / 2. * 1.05);
            }
        }
    }

    @Override
    public void readConfig(Reader reader) throws IOException {
        setTabId(reader.readChar());
        setPositionId(reader.readChar());
        setName(reader.readString());
        setTitle(reader.readString());
        setAxisX(readAxis(reader));
        setAxisY(readAxis(reader));
        getData().resize(getAxisX().getNbins() * 2);
    }

    @Override
    public void writeConfig(Writer writer) throws IOException {
        writer.writeString(getDataType());
        writer.writeChar((char) getTabId());
        writer.writeChar((char) getPositionId());
        writer.writeString(getName());
        writer.writeString(getTitle());
        writeAxis(writer, getAxisX());
        writeAxis(writer, getAxisY());
    }

    @Override
    public Histo clone() {
        Histo h = (Histo) HistoFactory.create(getDataType());
        h.setName(getName());
        h.setTitle(getTitle());
        h.getAxisX().copy(getAxisX());
        h.getAxisY().copy(getAxisY());
        h.getData().copy(getData());
        return h;
    }

    public void setNbinsx(int nbinsx) {
        axisX.setNbins(nbinsx);
        data.resize(nbinsx * 2);
    }

}
